core/drivers/sam/at91_audio_pll.c

```diff
@@ -0,0 +1,420 @@
+	io_clrsetbits32(apad_ck->base + AT91_PMC_AUDIO_PLL0,
+			AT91_PMC_AUDIO_PLL_PADEN, AT91_PMC_AUDIO_PLL_PADEN);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result clk_audio_pll_pmc_enable(struct clk *clk)
+{
+	struct clk_audio_pmc *apmc_ck = clk->priv;
+
+	io_clrsetbits32(apmc_ck->base + AT91_PMC_AUDIO_PLL0,
+			AT91_PMC_AUDIO_PLL_PMCEN |
+			AT91_PMC_AUDIO_PLL_QDPMC_MASK,
+			AT91_PMC_AUDIO_PLL_PMCEN |
+			AT91_PMC_AUDIO_PLL_QDPMC(apmc_ck->qdpmc));
+	return TEE_SUCCESS;
+}
+
+static void clk_audio_pll_frac_disable(struct clk *clk)
+{
+	struct clk_audio_frac *frac = clk->priv;
+
+	io_clrsetbits32(frac->base + AT91_PMC_AUDIO_PLL0,
+			AT91_PMC_AUDIO_PLL_PLLEN, 0);
+	/* do it in 2 separated writes */
+	io_clrsetbits32(frac->base + AT91_PMC_AUDIO_PLL0,
+			AT91_PMC_AUDIO_PLL_RESETN, 0);
+}
+
+static void clk_audio_pll_pad_disable(struct clk *clk)
+{
+	struct clk_audio_pad *apad_ck = clk->priv;
+
+	io_clrsetbits32(apad_ck->base + AT91_PMC_AUDIO_PLL0,
+			AT91_PMC_AUDIO_PLL_PADEN, 0);
+}
+
+static void clk_audio_pll_pmc_disable(struct clk *clk)
+{
+	struct clk_audio_pmc *apmc_ck = clk->priv;
+
+	io_clrsetbits32(apmc_ck->base + AT91_PMC_AUDIO_PLL0,
+			AT91_PMC_AUDIO_PLL_PMCEN, 0);
+}
+
+static unsigned long clk_audio_pll_fout(unsigned long parent_rate,
+					unsigned long nd, unsigned long fracr)
+{
+	unsigned long long fr = (unsigned long long)parent_rate * fracr;
+
+	DMSG("A PLL: fr = %llu", fr);
+
+	fr = UDIV_ROUND_NEAREST(fr, AUDIO_PLL_DIV_FRAC);
+
+	DMSG("A PLL: fr = %llu", fr);
+
+	return parent_rate * (nd + 1) + fr;
+}
+
+static unsigned long clk_audio_pll_frac_get_rate(struct clk *clk,
+						 unsigned long parent_rate)
+{
+	struct clk_audio_frac *frac = clk->priv;
+	unsigned long fout = 0;
+
+	fout = clk_audio_pll_fout(parent_rate, frac->nd, frac->fracr);
+
+	DMSG("A PLL: %s, fout = %lu (nd = %u, fracr = %lu)\n", __func__,
+	     fout, frac->nd, (unsigned long)frac->fracr);
+
+	return fout;
+}
+
+static unsigned long clk_audio_pll_pad_get_rate(struct clk *clk,
+						unsigned long parent_rate)
+{
+	struct clk_audio_pad *apad_ck = clk->priv;
+	unsigned long apad_rate = 0;
+
+	if (apad_ck->qdaudio && apad_ck->div)
+		apad_rate = parent_rate / (apad_ck->qdaudio * apad_ck->div);
+
+	DMSG("A PLL/PAD: %s, apad_rate = %lu (div = %u, qdaudio = %u)\n",
+	     __func__, apad_rate, apad_ck->div, apad_ck->qdaudio);
+
+	return apad_rate;
+}
+
+static unsigned long clk_audio_pll_pmc_get_rate(struct clk *clk,
+						unsigned long parent_rate)
+{
+	struct clk_audio_pmc *apmc_ck = clk->priv;
+	unsigned long apmc_rate = 0;
+
+	apmc_rate = parent_rate / (apmc_ck->qdpmc + 1);
+
+	DMSG("A PLL/PMC: %s, apmc_rate = %lu (qdpmc = %u)\n", __func__,
+	     apmc_rate, apmc_ck->qdpmc);
+
+	return apmc_rate;
+}
+
+static TEE_Result clk_audio_pll_frac_compute_frac(unsigned long rate,
+						  unsigned long parent_rate,
+						  unsigned long *nd,
+						  unsigned long *fracr)
+{
+	unsigned long long tmp = 0;
+	unsigned long long rem = 0;
+
+	if (!rate)
+		return -1;
```

Comment:
`TEE_ERROR_GENERIC` or another TEE_Result compliant code.

Reply:
used TEE_ERROR_BAD_PARAMETERS since rate is a param